### ISSUE TITLE
Getting plugin not found error

### ISSUE THREAD

**Issue by sworup** [openmsupply/msupply-dashboard-app#78]
# Details

```
t=2022-03-08T23:52:22+1300 lvl=eror msg="Could not start plugin" logger=plugin.manager pluginId=msupplyfoundation-datasource err="Unrecognized remote plugin message: 2022/03/08 23:52:22 notifying bugsnag: Binary was compiled with 'CGO_ENABLED=0', go-sqlite3 requires cgo to work. This is a stub\n\nThis usually means that the plugin is either invalid or simply needs to be recompiled to support the latest protocol."
```

In mac grafana tries to use the ARM version of the executable which also breaks saying could not find file.

I did recompiled the app in my m1 mac system but it was broken before I tinkered. I forgot to take the error log entry though.

Definitely getting this error because Grafana for some reasons cannot run the backend executable.

**Comment by sworup** [comment]
@joshxg any suggestions? or maybe you can recompile the executable to see if Grafana picks it up.

**Comment by joshxg** [comment]
There is a bugsnag for this:

![image](https://user-images.githubusercontent.com/35858975/157305045-5bda0e93-86ea-4f77-8c18-1f0b28135bf0.png)


(If you can see: https://app.bugsnag.com/sustainable-solutions-nz-ltd/msupply-dashboard-app-1/errors/62134cc6f577200008b1cac3?filters[event.since]=30d&filters[error.status]=open)

Which isn't particularly useful 😁 

Are you use an m1? It might be an issue with cgo not being enabled?  https://www.reddit.com/r/golang/comments/ka5ipb/gosqlite3_requires_cgo_to_work/

**Comment by sworup** [comment]
> Are you use an m1?

@joshxg yes I am using m1, which might not work locally for me, but the exe does get compiled. 
Also I tried compiling in the windows `demo-board` vm server and I get the same error with the exe generated from that version. So I kinda think m1 compiled executable exe might still work if we can get the Go code error free.
I will look into this `CGO_ENABLED` and will get back to you if I get stuck.
Thank you for the prompt reply.

**Comment by sworup** [comment]
I enabled CGO to compile the app and got another error:

```
PS C:\Users\Administrator\Documents\GitHub\msupply-dashboard-app\backend> go run mage.go
# runtime/cgo
gcc: error: unrecognized command line option '-marm'; did you mean '-mabm'?
# runtime/cgo
gcc: error: x86_64: No such file or directory
gcc: error: unrecognized command line option '-arch'
# runtime/cgo
gcc_linux_arm64.c: In function '_cgo_sys_thread_start':
gcc_linux_arm64.c:22:2: error: unknown type name 'sigset_t'; did you mean '_sigset_t'?
   22 |  sigset_t ign, oset;
      |  ^~~~~~~~
      |  _sigset_t
gcc_linux_arm64.c:27:2: error: implicit declaration of function 'sigfillset' [-Werror=implicit-function-declaration]
   27 |  sigfillset(&ign);
      |  ^~~~~~~~~~
gcc_linux_arm64.c:22:16: error: unused variable 'oset' [-Werror=unused-variable]
   22 |  sigset_t ign, oset;
      |                ^~~~
cc1.exe: all warnings being treated as errors
# runtime/cgo
gcc_linux_amd64.c: In function '_cgo_sys_thread_start':
gcc_linux_amd64.c:61:2: error: unknown type name 'sigset_t'; did you mean '_sigset_t'?
   61 |  sigset_t ign, oset;
      |  ^~~~~~~~
      |  _sigset_t
gcc_linux_amd64.c:66:2: error: implicit declaration of function 'sigfillset' [-Werror=implicit-function-declaration]
   66 |  sigfillset(&ign);
      |  ^~~~~~~~~~
gcc_linux_amd64.c:61:16: error: unused variable 'oset' [-Werror=unused-variable]
   61 |  sigset_t ign, oset;
      |                ^~~~
cc1.exe: all warnings being treated as errors
Error: running "go build -o dist\gpx_msupply-datasource_linux_arm -ldflags -w -s ./pkg" failed with exit code 2
running "go build -o dist\gpx_msupply-datasource_darwin_amd64 -ldflags -w -s ./pkg" failed with exit code 2
running "go build -o dist\gpx_msupply-datasource_linux_arm64 -ldflags -w -s ./pkg" failed with exit code 2
running "go build -o dist\gpx_msupply-datasource_linux_amd64 -ldflags -w -s ./pkg" failed with exit code 2
exit status 2
```

**Comment by sworup** [comment]
Ok I got the plugin to compile but now I am seeing this error:

```
logger=plugin.manager t=2022-03-17T01:16:06.72+1300 lvl=eror msg="Could not start plugin" pluginId=msupplyfoundation-datasource err="Unrecognized remote plugin message: 2022/03/17 01:16:06 notifying bugsnag: runtime error: index out of range [0] with length 0\n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol."
```

@joshxg how do see the bugsnag bug report? Maybe that has details.

Compile the backend with CGO did not do the trick. I had to do `go install github...sqlite3` with CGO_ENABLED = 1 and then run go run mage.go.

**Comment by joshxg** [comment]
Nice! Maybe that annoyance should be added to the README..

Bugsnag is here https://app.bugsnag.com/sustainable-solutions-nz-ltd/msupply-dashboard-app-1/timeline?filters[event.since]=30d&filters[error.status]=open&pivot_tab=error

I can't see anything with a description you've given.. only CGO problems in the last ~1 week.

**Comment by sworup** [comment]
The point I am at right now:
The source of the error:

```
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg="panic: runtime error: index out of range [0] with length 0"
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg=
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg="goroutine 1 [running]:"
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg="github.com/grafana/simple-datasource-backend/pkg/api.GetEmails({{0xc00029aee5, 0x5}, {0xc00029af10, 0xc}, {0xc0000b2a80, 0x23}}, {0x0, 0x0, 0xc0000b2c00}, 0xaf)"
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg="\t/Users/sworup/sussol/Github/dashboard-plugins/msupply-dashboard-app/backend/pkg/api/users.go:23 +0x6e5"
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg=github.com/grafana/simple-datasource-backend/pkg/reportEmailer.(*ReportEmailer).CreateReports(0xc00028e820)
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg="\t/Users/sworup/sussol/Github/dashboard-plugins/msupply-dashboard-app/backend/pkg/reportEmailer/reportEmailer.go:221 +0x5dd"
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg=main.main()
logger=plugin.loader pluginID=msupplyfoundation-datasource t=2022-03-22T20:19:12.28+1300 lvl=dbug msg="\t/Users/sworup/sussol/Github/dashboard-plugins/msupply-dashboard-app/backend/pkg/main.go:51 +0x249"
logger=plugin.manager t=2022-03-22T20:19:12.28+1300 lvl=eror msg="Could not start plugin" pluginId=msupplyfoundation-datasource err="Unrecognized remote plugin message: 2022/03/22 20:19:12 notifying bugsnag: runtime error: index out of range [0] with length 0\n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol."

```

**Comment by sworup** [comment]
@mark-prins 

@bijaySussol told me that you are not getting this error?

**Comment by mark-prins** [comment]
Nice work @sworup! 
I wasn't getting the error on the mac (m1.. which compiled ok for me) - but do see it on the demo server which is windows.

**Comment by sworup** [comment]
Hi @mark-prins
A few question:

- You said it compiled for you on mac. Did you have CGO enabled in mac? I had to install linux docker to build the executables. 
- I ended up finding a version of sqlite3 without CGO requiremnt, which I thought would be a convinient way.
- You said you saw the error in demo-dash, did you compile it in your computer and copy to plugin to demo-dash? Since, the plugin that was there, I compiled it yesterday, before reporting the error there. I wonder if the version you had compiled would work in demo-dash.
- Even if the compilation worked on mac, when I went in to the reports, I would get a no user error. Which is on par with the windows error, that would say user_Id[] array is out of bound on key 0. I think mac is also getting the same error as windows, its just that windows rejects it during compilation while mac supresses it.

**Comment by sworup** [comment]
> There is a bugsnag for this:
> 
> [image](https://user-images.githubusercontent.com/35858975/157305045-5bda0e93-86ea-4f77-8c18-1f0b28135bf0.png)
> 
> (If you can see: [https://app.bugsnag.com/sustainable-solutions-nz-ltd/msupply-dashboard-app-1/errors/62134cc6f577200008b1cac3?filters[event.since]=30d&filters[error.status]=open](https://app.bugsnag.com/sustainable-solutions-nz-ltd/msupply-dashboard-app-1/errors/62134cc6f577200008b1cac3?filters%5Bevent.since%5D=30d&filters%5Berror.status%5D=open))
> 
> Which isn't particularly useful 😁
> 
> Are you use an m1? It might be an issue with cgo not being enabled? https://www.reddit.com/r/golang/comments/ka5ipb/gosqlite3_requires_cgo_to_work/

The error josh mentioned is because the database path (which is hardcoded) could not be resolved. Normally the path is ../data/msupply.db, which looks for the db in the data folder above the plugin folder. Since the dev environment has plugin path set custom, this data directory does not exist. This directory structure works for demo-dash windows server. 

In mac I generally change this path to the data directory of grafana, for instance in my installation it was `var/lib/grafana` so the path would be `var/lib/grafana/msupply.db`.

**Comment by mark-prins** [comment]
> Did you have CGO enabled in mac? 

Not knowingly - I haven't installed CGO.. it just works for me. Can I check to see if I have GO enabled?

> did you compile it in your computer and copy to plugin to demo-dash? 

The back end plugin was compiled on demo-dash server, I've been using that to compile the windows binary. Are you able to compile windows binaries on mac?

> Normally the path is ../data/msupply.db, which looks for the db in the data folder above the plugin folder

Yeah, that was a bit of a pain when running locally. I just created a `data` folder a level up on my mac and put the db in there. Wasn't able to reference an absolute path when trying to load the database unfortunately.

**Comment by sworup** [comment]
@mark-prins 

Here's the reason why the errors were showing up and what I did to fix them with #80 

- CGO issue: The Go executable of the plugin was refusing to load because go-sqlite3 needed the executable to be built with CGO_ENABLED=1. Normally when you build GO code, it builds with CGO_ENABLED=0.
  - Fix: I managed to work up enabling CGO in demo-dash sever and even in my docker Linux environment but it is too much of a hassle to set up correctly. I found a version of sqlite3 that was CGO free [modernc.org/sqlite plugin](https://github.com/openmsupply/msupply-dashboard-app/pull/80/commits/0e208bec7aab57db130761e4d06041c186789305) and replaced the package in our code. That made things easier as I could build with CGO_ENABLED=0
- After that fix I had emailport being an string when int was expected.
  - Turns out in my other fix for settings, I had mistakenly saved emailPort as emailport. So now the response was sending both, and the system was looking at emailport. Fixed it by renaming the variable. [f53b79b](https://github.com/openmsupply/msupply-dashboard-app/pull/80/commits/f53b79b450b02806c1ba5c79a9a9e867ea967e34)
- After that the system wanted to load a single scheduler expecting a scheduler resorve. Added that and that was fixed. [e98fb24](https://github.com/openmsupply/msupply-dashboard-app/pull/80/commits/e98fb2429ad8f9b53de057e33c41cbc419d4edec)
- A minor bug was the frontend filter to filter user in Report group page was bugging out, it wasn't filtering correctly.
  - After some head scrating, found that the li component loop expects an unique key, the key was set as name field, which could have duplicate values. I set id as the key and that fixed it [5c64472](https://github.com/openmsupply/msupply-dashboard-app/pull/80/commits/5c6447232d31abad6f9d29490dc91bf3c1ed2e81)
- Main error: The main error was user array out of bound at 0. The reason this error was showing was because one of the scheduler had a report group selected that did not have any users. Since the users number is zero it was erroring out as our go code did not have code to handle this situation. In normal flow this would have stilled errored out but silently. But since this happened during the initialisation phase, while trying to load pre-existing schedulers, Grafana just said, "hay this plugin executable has error, I am not going to load it."
  - I check the case where users in users group is 0, and just added a break so that codes after that would not try to run. That temperorily fixes it but I think there are graceful ways to do this.

**Comment by Mai21** [comment]
Thanks, @sworup 

> Main error: The main error was user array out of bound at 0. The reason this error was showing was because one of the schedulers had a report group selected that did not have any users. Since the users number is zero it was erroring out as our go code did not have code to handle this situation. 

This information helped me to fix this error in Windows server 2016. It was shown when we restarted Grafana and "Plugin unavailable" message showed when we accessed the plugin setting pages.

```
t=2022-04-01T04:32:19+0100 lvl=eror msg="Failed to start plugin" logger=plugins.backend pluginId=msupplyfoundation-datasource error="Unrecognized remote plugin message: 2022/04/01 04:32:19 notifying bugsnag: runtime error: index out of range [0] with length 0\n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol."
```

It was not a permanent solution but inserting a record manually into ReportGroupMembership table enabled me to use the scheduler report anyway. 

I really appreciate your bug fix in #80 !
